File: LeetCode/155. Min Stack.java
class MinStack {
    LinkedList<Node> st = new LinkedList<>();

    public MinStack() {} //SC: O(count of elements)
    
    public void push(int val) {
        if (!st.isEmpty()) {
            st.push(new Node(val, st.peek().minVal));
        } else {
            st.push(new Node(val, Integer.MAX_VALUE));
        }
    } //TC: O(1)
    
    public void pop() {
        st.poll();
    } //TC: O(1)
    
    public int top() {
        return st.peek().val;
    } //TC: O(1)
    
    public int getMin() {
        return st.peek().minVal;
    } //TC: O(1)
}
class Node {
    int minVal = Integer.MAX_VALUE;
    int val = 0;
    Node(int val, int lastMin) {
        this.val = val;
        this.minVal = Math.min(lastMin, this.val);
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: LeetCode/155. Min Stack.kt
class MinStack() {
    private val st = LinkedList<Int>()
    private val minSt = LinkedList<Int>() //aside from regular stack, setup a stack to store min values

    fun push(`val`: Int)  = st.push(`val`).also {
        val min = minSt.peek()?.let {
            Math.min(`val`, it)
        } ?: `val` //min = val if min stack is empty, otherwise take the less of peeked min value or val
        minSt.push(min) //when pushing a value to regular stack, also push a min value to min stack
    }

    fun pop() {
        st.pop()
        minSt.pop() //pop min stack along with regular stack
    }

    fun top() = st.peek()

    fun getMin() = minSt.peek() //peek the min stack for current min value

}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = MinStack()
 * obj.push(`val`)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
